correcting #17940 to get rid of conditionset for simple exp equation

MaqOwais opened the pull request:
correcting #17940 to get rid of conditionset for simple exp equation

```
INPUT:
solveset(exp(exp(x))-5, x)

AFTER MODIFICATION
OUTPUT:
ImageSet(Lambda(_n, 2*_n*I*pi + log(log(5))), Integers)

```
<!-- BEGIN RELEASE NOTES -->
NO ENTRY
<!-- END RELEASE NOTES -->

MaqOwais committed:
correcting #17940

MaqOwais committed:
Test case for solveset

oscarbenjamin commented:
Shouldn't the solution be
```julia
In [5]: x = 2*n*I*pi + log(log(5) + 2*m*pi*I)                                                                                                                 

In [6]: x                                                                                                                                                     
Out[6]: 2⋅ⅈ⋅π⋅n + log(2⋅ⅈ⋅π⋅m + log(5))

In [7]: exp(exp(x))-5                                                                                                                                         
Out[7]: 0
```

smichr commented on sympy/solvers/solveset.py:
```diff
@@ -327,7 +327,12 @@ def _invert_complex(f, g_ys, symbol):
-        if isinstance(g_ys, FiniteSet):
+        x = Symbol('x')
+        if (len((f.args)) < 2) and (f.args == (exp(x),)):
+            print(len((f.args)))
```
you will want to remove the `print` before committing

MaqOwais replied:
yeah ,
my apologies

MaqOwais commented:
> Shouldn't the solution be

Yes , that's correct

MaqOwais commented:
@oscarbenjamin 
Any suggestions !

MaqOwais committed:
Update solveset.py

MaqOwais commented:
I think imageset do not support for multi variables ...
like @oscarbenjamin suggested the solution as  2⋅ⅈ⋅π⋅n + log(2⋅ⅈ⋅π⋅m + log(5)) 
 where we see m , n as different two variables ( arbitrary constants ) ..

oscarbenjamin commented:
Here are 3 different ways to represent that set:
```julia
In [6]: ImageSet(Lambda((n, m), 2*n*pi*I + log(2*m*pi*I + log(5))), Integers, Integers)                                                        
Out[6]: {2⋅ⅈ⋅π⋅n + log(2⋅ⅈ⋅π⋅m + log(5)) | n ∊ ℤ , m ∊ ℤ}

In [7]: ImageSet(Lambda(((n, m),), 2*n*pi*I + log(2*m*pi*I + log(5))), Integers**2)                                                            
Out[7]: 
⎧                                            2⎫
⎨2⋅ⅈ⋅π⋅n + log(2⋅ⅈ⋅π⋅m + log(5)) | (n, m) ∊ ℤ ⎬
⎩                                             ⎭

In [8]: ImageSet(Lambda((n, x), 2*m*pi*I+log(x)), Integers, ImageSet(Lambda(m, 2*I*m*pi + log(5)), Integers))                                  
Out[8]: {2⋅ⅈ⋅π⋅m + log(x) | n ∊ ℤ , x ∊ {2⋅ⅈ⋅π⋅m + log(5) | m ∊ ℤ}}
```

MaqOwais commented:
Even though the lines of code are change and it's returning the required output but it's raising assertion error
```
This is returning `False`
solveset(exp(exp(x))-5, x) == ImageSet(Lambda(((k, n),), I*(2*k*pi + arg(2*n*I*pi + log(5))) + log(Abs(2*n*I*pi + log(5)))), ProductSet(Integers, Integers))
```
I've tried to change by putting _k or _n and _m or _n 
```
line 334 - 339
    if isinstance(f, exp):
        if isinstance(g_ys, FiniteSet):
            exp_invs = Union(*[imageset(Lambda(n, I*(2*n*pi + arg(g_y)) +
                                               log(Abs(g_y))), S.Integers)
                               for g_y in g_ys if g_y != 0])
            return _invert_complex(f.args[0], exp_invs, symbol)


```

{This could be only special case for `exp(exp(x)) - k `}
MODIFIED VERSION
```
    if isinstance(f, exp):
        x = Symbol('x')
        if isinstance(g_ys,list) and not isinstance(g_ys, FiniteSet) \
                        and not isinstance(g_ys,ImageSet):
            k = Dummy('k')
            exp_invs = Union(*[imageset(Lambda(((k, n),), I*(2*k*pi) + 
                                log(Abs(g_y))), S.Integers**2)])
            return _invert_complex(f.args[0], exp_invs,symbol)

        elif isinstance(g_ys, FiniteSet):
            fg = [ I*(2*n*pi + arg(g_y)) + log(Abs(g_y)) for g_y in g_ys if g_y != 0]
            exp_invs = Union(*[imageset(Lambda(n, I*(2*n*pi + arg(g_y)) +
                                            log(Abs(g_y))), S.Integers)
                            for g_y in g_ys if g_y != 0])
            if f.args[0] != x :
                return _invert_complex(f.args[0], fg , symbol)    
            return _invert_complex(f.args[0], exp_invs, symbol)

Modified Output:
ImageSet(Lambda(((_k, _n),), 2*_k*I*pi + log(Abs(2*_n*I*pi + log(5)))), ProductSet(Integers, Integers))

```

Please have a view on it.

oscarbenjamin commented:
You probably need to use `dummy_eq`.

MaqOwais commented:
```
This code is also returning `False`

solveset(exp(exp(x))-5, x).dummy_eq(ImageSet(Lambda(((_k, _n),), 2*_k*I*pi + log(Abs(2*_n*I*pi + log(5)))), ProductSet(Integers, Integers)))
```

oscarbenjamin commented:
Can you show the output of using `srepr` on the two expression you are comparing?

MaqOwais commented:
```
p1 = solveset(exp(exp(x))-5, x)
srepr(p1)
"ImageSet(Lambda(Tuple(Tuple(Dummy('k', dummy_index=2914607), Dummy('n', dummy_index=2914606))), Add(Mul(Integer(2), Dummy('k', dummy_index=2914607), I, pi), log(Abs(Add(Mul(Integer(2), Dummy('n', dummy_index=2914604), I, pi), log(Integer(5))))))), ProductSet(Integers, Integers))"

```

```
p2 = ImageSet(Lambda(((_k, _n),), 2*_k*I*pi + log(Abs(2*_n*I*pi + log(5)))), ProductSet(Integers, Integers))
srepr(p2)
"ImageSet(Lambda(Tuple(Tuple(Dummy('k', dummy_index=2914600), Dummy('n', dummy_index=2914601))), Add(Mul(Integer(2), Dummy('k', dummy_index=2914600), I, pi), log(Abs(Add(Mul(Integer(2), Dummy('n', dummy_index=2914601), I, pi), log(Integer(5))))))), ProductSet(Integers, Integers))"
```

czgdp1807 commented on sympy/solvers/tests/test_solveset.py:
```diff
@@ -1008,6 +1008,9 @@ def test_solveset():
                                                   S.Integers)
+    # ISSUE 17940
+    _n = Dummy('n')
+    assert solveset(exp(exp(x))-5,x) == ImageSet(Lambda(_n, 2*_n*I*pi + log(log(5))), S.Integers)
```
```suggestion
    assert solveset(exp(exp(x)) - 5, x) == ImageSet(Lambda(_n, 2*_n*I*pi + log(log(5))), S.Integers)
```

MaqOwais replied:
Thanks a lot for suggestion and response.
:)

I think this should be discussed more, here the return type is different from all other similar problems.

oscarbenjamin commented:
Comparing p1 and p2 I see that one of them has multiple dummies for `n`:
```
In [84]: p2.atoms(Symbol)                                                                                                         
Out[84]: {k, n}

In [87]: p1.atoms(Symbol)                                                                                                         
Out[87]: {k, n, n}
```
This is because in p1 the symbol n in the lambda signature is a different `Dummy` from the symbol n in the lambda expression.
```
In [149]: p1.args[0]                                                                                                              
Out[149]: ((k, n),) ↦ 2⋅k⋅ⅈ⋅π + log(│2⋅n⋅ⅈ⋅π + log(5)│)

In [150]: p1.args[0]((0, 0))                                                                                                      
Out[150]: log(│2⋅n⋅ⅈ⋅π + log(5)│)

In [151]: p2.args[0]((0, 0))                                                                                                      
Out[151]: log(log(5))
```

MaqOwais commented:
Yeah that's right but I think it can't be avoided, since I tried any other ways to arrive at solution but I couldn't get any clue ...
can you please give any suggestions

czgdp1807 commented:
@MaqOwais Any news on this?

czgdp1807 commented:
Closing this in favor of https://github.com/sympy/sympy/pull/20729
